R/io.R

```diff
@@ -297,36 +297,58 @@ reclassify_raster <- function(crop_cover_levels){
-
-  reclass_table_df <- dplyr::rename(reclass_table, CDL_ID = ID,
-                                            GCAM_Class = GCAM_commodity)
+  # First link uses the GCAM crop table and links the CDL table by similar crop names. 40 IDs get filled by this operation.
+  gcam_csv %>%
+    dplyr::select(item, GTAP_crop, GCAM_commodity, IFA2002_crop) %>%
+    left_join(cdl_classes, by = c("GTAP_crop" = "CDL_Class")) %>%
+    left_join(cdl_classes, by = c("IFA2002_crop" = "CDL_Class")) %>%
+    mutate(ID = case_when(
+      is.na(ID.x) & is.na(ID.y) ~ NA_integer_,
+      is.na(ID.x) & !is.na(ID.y) ~ ID.y,
+      !is.na(ID.x) & is.na(ID.y) ~ ID.x,
+      !is.na(ID.x) & !is.na(ID.y) ~ ID.x)) %>%
+    left_join(gcam_classes, by = "GCAM_commodity") %>%
+    dplyr::select(ID, GCAM_ID, item) %>%
+    left_join(cdl_classes, gcam_csv, by = "ID") -> gcam_link
+
+  # This is then merged with the CDL IDs. All of the IDs that are still NA are made into a new table to do the second link.
+  crop_cover_levels %>%
+    left_join(gcam_link, by = "ID") %>%
+    left_join(gcam_classes, by = "GCAM_ID") %>%
+    dplyr::select(ID, Class_Names, GCAM_ID, CDL_Class, GCAM_commodity) -> crop_level_join
+
+  crop_level_join[is.na(crop_level_join$GCAM_commodity), ] -> na_crops
+
+  # This second link uses grepl to assign crops that are still NA to GCAM crops.
+  na_crops %>%
+    mutate(GCAM_Commodity = dplyr::case_when(
+      grepl("Oats", Class_Names) ~ "OtherGrain",
+      grepl("Grains", Class_Names) | grepl("Other Small Grains", Class_Names) ~ "OtherGrain",
+      Class_Names %in% c("Corn", "Sweet Corn", "Pop or Orn Corn") ~ "Corn",
+      Class_Names %in% c("Sod/Grass Seed", "Switchgrass") ~ "FodderGrass", Class_Names %in% c("Canola", "Rape Seed", "Camelina") ~ "OilCrop",
+      Class_Names %in% c("Other Hay/Non Alfalfa", "Clover/Wildflowers", "Vetch") ~ "FodderHerb", grepl("Wheat", Class_Names) ~ "Wheat",
+      grepl("Speltz", Class_Names) ~ "Wheat",grepl("Sweet Potatoes", Class_Names) ~ "Root_Tuber",
+      grepl("Sugar", Class_Names) ~ "SugarCrop", grepl("Flaxseed", Class_Names) ~ "FiberCrop", grepl("Dbl Crop", Class_Names) ~ "MiscCrop",
+      grepl("Developed", Class_Names) | grepl("Water", Class_Names) | grepl("Perennial", Class_Names) |
+        grepl("Barren", Class_Names) | grepl("Forest", Class_Names) | grepl("land", Class_Names) | grepl("Background", Class_Names) |
+        grepl("Clouds", Class_Names) | grepl("Undefined", Class_Names) | grepl("Aqua", Class_Names) ~ NA_character_, TRUE ~ "MiscCrop")) %>%
+    dplyr::select(ID, Class_Names, GCAM_Commodity) -> na_crops_class
+
+  # This joins both tables together so that every CDL crop is assigned a GCAM crop
+  left_join(crop_level_join, na_crops_class, by = "ID") -> full_crop_table
+
+  # FIlls in NA of GCAM_commodity
+  ifelse(is.na(full_crop_table$GCAM_commodity), full_crop_table$GCAM_Commodity, full_crop_table$GCAM_commodity) -> full_crop_table$GCAM_commodity
```

Comment:
If you want to neaten this up then you can use `mutate(GCAM_commodity = if_else(is.na(GCAM_commodity), GCAM_Commodity, GCAM_commodity))` Clear col names would be useful (took me a while to realize the distinction was just the capital C).